What happened to UUIDv2? (#463)
unicorn-utterances/unicorn-utterances: content/blog/what-happened-to-uuid-v2/index.md

---
{
    title: "What Happened to UUIDv2?",
    description: "You may have heard that "UUIDv2 is bad". Maybe you've never even heard of them and thought UUIDs went from v1 to v3. Why is that? What did UUIDv2 do wrong?",
    published: '2023-01-15T21:52:59.284Z',
    authors: ['crutchcorn'],
    tags: ['computer science'],
    attached: [],
    license: 'cc-by-4',
    series: "Explaining UUIDs",
    order: 2
}
---

If you read [my last post introducing UUIDs](/posts/what-are-uuids), you'll remember that I said:

> UUIDv2 is _weird_.

After this, [I left a very short (one or two sentences) explanation of what a UUIDv2 is](/posts/what-are-uuids#UUIDv2).

Why would I do that [when UUIDv3 and UUIDv5 got four whole headers worth of explanation](/posts/what-are-uuids#UUIDv3and5)? What makes UUIDv2 different or "weird," as I put it?

The answer is that while UUID _is_ a form of UUID by some definitions, it effectively is not in others.

> What?!

> I know... It's a bit jarring. What's worse, even within the definitions we _do_ have, UUID is deeply flawed. Enough so that most are comfortable leaving it in the past.

Reviewer: ```suggestion
I know... It's a bit jarring. But, even within the definitions we have, UUID is deeply flawed, enough so that most are comfortable leaving it in the past.
```

> I know... It's a bit jarring. What's worse, even within the definitions we _do_ have, UUID is deeply flawed. Enough so that most are comfortable leaving it in the past.

Reviewer: is it "UUID is deeply flawed?" or "UUIDvs2 is..."?

To figure out how UUIDv2 became this way, let's explore the following:

- [What makes a UUID a UUID?](#specs)
- [How is UUIDv2 defined?](#uuidv2-definition)
- [What was UUIDv2 supposed to be used for?](#posix-data)
- [What are the problems with UUIDv2?](#problems-with-uuidv2)

# What makes a UUID a UUID? {#specs}

> As you may have noticed, UUIDs aren't a free-for-all. There are specific rules that apply to UUID that come from a document released in 2005 by the "[Internet Engineering Task Force](https://www.ietf.org/)" (IETF). This document is known as [`RFC 4122`](https://datatracker.ietf.org/doc/html/rfc4122) and acts as the de-facto UUID specification.

Reviewer: "free-for-all" isn't as common as other expressions, a good majority won't get what this means

> Description: DCE Security version, with embedded POSIX UIDs.

That's it.

Even in a proposal to update RFC 4122	

This makes the UUIDv2 the only official version of UUID that does not have an explicit definition as part of the 2005 specification.

This doesn't mean that it doesn't have rules, however; The UUIDv2 specification can be found in an older standard set for UUIDs. This older standard was created by an organizational body then called the "Open Software Foundation" (OSF), now called "The Open Group."

This older specification, which outlined the rules for UUIDv2 and was released in 1997, is called [the DCE 1.1 Authentication and Security Services specification](https://pubs.opengroup.org/onlinepubs/9696989899/toc.htm).

> Interesting in more history of UUIDs? [Twilio Segment's blog has an amazing history lesson about how they came to be](https://segment.com/blog/a-brief-history-of-the-uuid/). 

# How is UUIDv2 Defined? {#uuidv2-definition}

While the DCE specification is happy to go in-depth about [the fine-grained details of UUIDv2](https://pubs.opengroup.org/onlinepubs/9696989899/chap5.htm#tagcjh_08_02_01_01), let's take a more zoomed-out look at it:

> UUIDv2 is almost to [UUIDv1](/posts/what-are-uuids#UUIDv1). It contains most of the same components:

Reviewer: missing word: "UUIDvs is almost --- to UUIDv1"

- A timestamp
- A version
- A variant
- A clock sequence
- A Mac address

However, there are a few small differences. Namely:

- The "Clock Sequence" from UUIDv1 is changed from 3 characters to 1 as the last 2 characters are replaced with a new "Local Domain" enum value.

- The "Low Time" is replaced with a "Local Domain Number."

> ![// TODO: Write](./UUIDv2.svg)

Reviewer: Don't forget this

**This was UUIDv2's original purpose: Encoding of POSIX data in a unique ID**: There are application instances where having the user's operating system information easily accessible would be handy in the resource's ID.

# Why do UUIDv2s suck? {#problems-with-uuidv2}

UUIDv2 has a lot of problems today:

- **Few implementations**: Because of the scarcity of UUIDv2 usage and the lack of formal specification in RFC 4122, there are very few implementations of UUIDv2 in most languages and libraries. This may make implementing them more challenging than other versions of UUID.
- **Difficult to research**: Because there are few implementations of UUIDv2, it's challenging to learn about this version of UUIDs. Most articles ([including my own introduction to UUIDs](/posts/what-are-uuids)) leave a single paragraph (if even!) about the subject.

Both of these feel a bit more like symptoms of a deeper-rooted problem. Dig deep enough, and you'll end up finding what this problem is: **UUIDv2 has a very high likelihood of ID collision**.

> This means that if you run UUIDv2 multiple times in rapid succession, the likelihood you'll get the exact same ID.

Reviewer: ```suggestion
This means that if you run UUIDv2 multiple times in rapid succession, you're very likely to get the exact same ID.
```

## Explaining How UUIDv1 Handles Dates

Let's look back at the previous article to see how UUIDv1 is structured:

![A UUID broken down into "Low Time", a dash, "Mid Time", a dash, "Version", "High Time", a dash, "Variant", "Clock", a dash, and finally a "MAC Address". An example UUIDv1 might be "4e2b4d4c-92e8-11ed-86a8-3fdb0085247e"](../what-are-uuids/UUIDv1.svg)

Here, we use a timestamp of 15 byes. Let's take the example UUIDv1 from that image:

```
4e2b4d4c-92e8-11ed-86a8-3fdb0085247e
```

And strip out just the `time` data:

```
4e2b4d4c-92e8-1ed
```

Now, we can sort them from `High Time`, `Mid Time`, then `Low Time`:

```
1ed-92e8-4e2b4d4c
```

Let's remove the dashes, leaving us with the following:

```
1ed92e84e2b4d4c
```

Now we can convert this [from a hex number to a decimal number](/posts/non-decimal-numbers-in-tech/):

```
138928689959882060
```

> What do we do with this number?

Well, this is a timestamp. Namely, this is the number of 100 nanoseconds intervals since October 15, 1582 at midnight UTC.

> This date may seem arbitrary, but this is the date of Gregorian reform to the Christian calendar.

> Let's use [`dayjs`](https://day.js.org) to see what this equates to in 

Reviewer: ```suggestion
Let's use [`dayjs`](https://day.js.org) to see what this equates to:
```
